src/com/haskforce/codeInsight/HaskellCompletionContributor.java

```diff
@@ -71,17 +72,32 @@ public void addCompletions(@NotNull CompletionParameters parameters,
                         // whether or not we were in the appropriate context.  This is useful to determine if following
                         // completions should be added.
                         completeKeywordImport(position, result);
+                        completeHaskellKeywords(position, result);
                         completeKeywordQualified(position, result);
                         if (completePragma(position, cacheHolder, result)) return;
                         if (completeModuleImport(position, cacheHolder, result)) return;
                         if (completeQualifiedNames(position, imports, cacheHolder, result)) return;
                         if (completeNameImport(position, cacheHolder, result)) return;
                         completeLocalNames(position, imports, cacheHolder, result);
+                        completeFunctionLocalNames(position,result);
                     }
                 }
         );
     }
 
+    /**
+     * Will only propose the haskell keywords when inside a haskell body. Will not take into account the specific position
+     * yet, just make sure it's inside the body, so as to not propose keywords other than import in imports
+     */
+    public static void completeHaskellKeywords(@NotNull final PsiElement position, @NotNull final CompletionResultSet result) {
+        if (HaskellUtil.isInsideBody(position)){
+            Set<String> haskellKeywords = HaskellNamesValidator.HASKELL_KEYWORDS;
+            for (String haskellKeyword : haskellKeywords) {
+                result.addElement(LookupElementBuilder.create(haskellKeyword));
+            }
+        }
+    }
+
```

Comment:
I agree with you. I just wanted to add a stop-gap for the rather annoying situation where I for example type d-o-enter and there's all of a sudden deepOf there when I have imported the lens library. It's on the little road map in my head to revisit this and make it behave correctly, and for me it's less annoying to have too much stuff in that selection list than having the wrong stuff in there.

That said, maybe it was the wrong approach here, maybe I should have limited my self to add only 'do' to the list and only when we're somewhere inside a FunOrPatdecl or so. If I re-read your remark, I think that's what your suggesting, isn't it? I change it?